module/extension/Director/tests/self_intersection.cpp

```diff
@@ -0,0 +1,141 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2022 NUbots <nubots@nubots.net>
+ */
+
+#include <catch2/catch_test_macros.hpp>
+#include <fmt/format.h>
+#include <nuclear>
+
+#include "Director.hpp"
+#include "TestBase.hpp"
+#include "util/diff_string.hpp"
+
+// Anonymous namespace to avoid name collisions
+namespace {
+
+    struct MainTask {};
+
+    template <int I>
+    struct SubTask {
+        SubTask() : subtask_id(I) {}
+        int subtask_id;
+    };
+
+    struct Dependency {
+        Dependency(const int& subtask_id_) : subtask_id(subtask_id_) {}
+        int subtask_id;
+    };
+
+    std::vector<std::string> events;
+
+    class TestReactor : public TestBase<TestReactor> {
+    public:
+        explicit TestReactor(std::unique_ptr<NUClear::Environment> environment)
+            : TestBase<TestReactor>(std::move(environment)) {
+
+            on<Provide<MainTask>>().then([this] {
+                // Emit optional tasks with one blocking the other
+                events.push_back("emitting task 1");
+                emit<Task>(std::make_unique<SubTask<1>>(), 0, true);
+                events.push_back("emitting task 2");
+                emit<Task>(std::make_unique<SubTask<2>>(), 0, true);
+            });
+
+            on<Provide<SubTask<1>>, Needs<Dependency>>().then([this](const SubTask<1>& t) {
+                events.push_back(fmt::format("subtask {} executed", t.subtask_id));
+
+                // Emit the dependency task
+                events.push_back(fmt::format("emitting dependency from subtask {}", t.subtask_id));
+                emit<Task>(std::make_unique<Dependency>(t.subtask_id));
+            });
+
+            on<Provide<SubTask<2>>, Needs<Dependency>>().then([this](const SubTask<2>& t) {
+                events.push_back(fmt::format("subtask {} executed", t.subtask_id));
+
+                // Emit the dependency task
+                events.push_back(fmt::format("emitting dependency from subtask {}", t.subtask_id));
+                emit<Task>(std::make_unique<Dependency>(t.subtask_id));
+            });
+
+            on<Provide<Dependency>>().then([this](const Dependency& t) {
+                events.push_back(fmt::format("dependency from subtask {}", t.subtask_id));
+            });
+
+            /**************
+             * TEST STEPS *
+             **************/
+            on<Trigger<Step<1>>, Priority::LOW>().then([this] {
+                events.push_back("emitting main task");
+                emit<Task>(std::make_unique<MainTask>());
+            });
+            on<Trigger<Step<2>>, Priority::LOW>().then([this] {
+                events.push_back("removing main task");
+                emit<Task>(std::unique_ptr<MainTask>());
```

Comment:
```suggestion
                emit<Task>(std::unique_ptr<MainTask>(nullptr));
```

Reply:
These do the same thing but I'll change it if you think it makes it clearer